scheme/sevsnp/evidence_handler.go

```diff
@@ -0,0 +1,530 @@
+func (o EvidenceHandler) GetSupportedMediaTypes() []string {
+	return EvidenceMediaTypes
+}
+
+func transformEvidenceToCorim(token *proto.AttestationToken) (*corim.UnsignedCorim, error) {
+	tsm, err := parseAttestationToken(token)
+	if err != nil {
+		return nil, err
+	}
+
+	reportProto, err := abi.ReportToProto(tsm.OutBlob)
+	if err != nil {
+		return nil, err
+	}
+
+	refValComid, err := sevsnpParser.ReportToComid(reportProto, 0)
+	if err != nil {
+		return nil, err
+	}
+
+	err = refValComid.Valid()
+	if err != nil {
+		return nil, err
+	}
+
+	refValCorim := corim.UnsignedCorim{}
+	refValCorim.SetProfile(EndorsementMediaTypeRV)
+	refValCorim.AddComid(refValComid)
+
+	return &refValCorim, nil
+}
+
+// ExtractClaims converts evidence in tsm-report format to our
+// "internal representation", which is in CoRIM format.
+func (o EvidenceHandler) ExtractClaims(
+	token *proto.AttestationToken,
+	_ []string,
+) (map[string]interface{}, error) {
+	var claimsSet map[string]interface{}
+
+	refValCorim, err := transformEvidenceToCorim(token)
+	if err != nil {
+		return nil, err
+	}
+
+	refValJson, err := refValCorim.ToJSON()
+	if err != nil {
+		return nil, err
+	}
+
+	err = json.Unmarshal(refValJson, &claimsSet)
+	if err != nil {
+		return nil, err
+	}
+
+	return claimsSet, nil
+}
+
+func extractProvisionedTA(trustAnchors []string) (*comid.CryptoKey, error) {
+	var (
+		taEndorsement *handler.Endorsement
+		avk           comid.KeyTriple
+	)
+
+	for i, t := range trustAnchors {
+		var endorsement handler.Endorsement
+
+		if err := json.Unmarshal([]byte(t), &endorsement); err != nil {
+			return nil, fmt.Errorf("could not decode endorsement at index %d: %w", i, err)
+		}
+
+		if endorsement.Type == handler.EndorsementType_VERIFICATION_KEY {
+			taEndorsement = &endorsement
+			break
+		}
+	}
+
+	if taEndorsement == nil {
+		return nil, ErrNoProvisionedTA
+	}
+
+	err := json.Unmarshal(taEndorsement.Attributes, &avk)
+	if err != nil {
+		return nil, err
+	}
+
+	// The StoreHandler takes care of ensuring that only one TA is
+	// supplied, we don't have to re-check it here.
+	provisionedArk := avk.VerifKeys[0]
+
+	return provisionedArk, nil
+}
+
+func validateCertificateChain(certChain *sevsnp.CertificateChain) error {
+	if len(certChain.GetArkCert()) == 0 {
+		return ErrNoARK
+	}
+
+	if len(certChain.GetAskCert()) == 0 {
+		return ErrNoASK
+	}
+
+	if len(certChain.GetVcekCert()) == 0 && len(certChain.GetVlekCert()) == 0 {
+		return ErrNoVEK
+	}
+
+	return nil
+}
+
+func validateTA(certChain *sevsnp.CertificateChain, provisionedArk *comid.CryptoKey) error {
+	if !bytes.Equal(certChain.GetArkCert(), []byte(provisionedArk.String())) {
+		return ErrTAMismatch
+	}
+
+	return nil
+}
+
+func validateReportIntegrity(tsm *tokens.TSMReport, certChain *sevsnp.CertificateChain) error {
+	var (
+		ark, ask, vcek, vlek []byte
+		attestation          sevsnp.Attestation
+	)
+
+	options := verify.Options{
```

Comment:
IIUC, CRL checking is controlled by `CheckRevocations` and is disabled by default.

Would it be worth checking?

Reply:
Thanks for pointing it out, @thomas-fossati!

We would need a caching server to accomplish this effectively. That's because, AMD's KDS will rate-limit repeated requests for CRL.